CI: Integrate macOS/arm64

ChinYikMing opened the pull request:
<div id='description'>
<h3>Summary by Bito</h3>
This PR enhances the CI pipeline support for macOS/arm64 builds by updating build configurations and managing macOS-specific dependencies. Key improvements include restructuring prebuilt artifact handling, implementing conditional compilation guards for RISC-V instructions, and overall optimizations for cross-platform compatibility.
<br>
<br>
<b>Unit tests added</b>: False
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 2
</div>

ChinYikMing committed:
Enable gcc-14 build rv_histogram on macOS/arm64

Error:
Undefined symbols for architecture arm64:
  "_ebreak_handler", referenced from:
      lC0 in riscv.o
  "_ecall_handler", referenced from:
      lC0 in riscv.o
  "_log_impl", referenced from:
      _rv_create in riscv.o
      _rv_create in riscv.o
      _rv_create in riscv.o
      _rv_profile in riscv.o
      _rv_profile in riscv.o
      _rv_run in riscv.o
  "_log_level_string", referenced from:
      _rv_create in riscv.o
  "_log_set_level", referenced from:
      _rv_create in riscv.o
  "_log_set_stdout_stream", referenced from:
      _rv_remap_stdstream in riscv.o
  "_memcpy_handler", referenced from:
      lC0 in riscv.o
  "_memory_delete", referenced from:
      _rv_delete in riscv.o
      _rv_create in riscv.o
  "_memory_ifetch", referenced from:
      _on_mem_ifetch in riscv.o
  "_memory_new", referenced from:
      _rv_create in riscv.o
  "_memory_read_b", referenced from:
      _on_mem_read_b in riscv.o
  "_memory_read_s", referenced from:
      _on_mem_read_s in riscv.o
  "_memory_read_w", referenced from:
      _on_mem_read_w in riscv.o
  "_memory_write_b", referenced from:
      _on_mem_write_b in riscv.o
  "_memory_write_s", referenced from:
      _on_mem_write_s in riscv.o
  "_memory_write_w", referenced from:
      _on_mem_write_w in riscv.o
  "_memset_handler", referenced from:
      lC0 in riscv.o
  "_rv", referenced from:
      _rv_async_block_clear in riscv.o
  "_rv_step", referenced from:
      _rv_run in riscv.o
      _rv_run in riscv.o
  "_trap_handler", referenced from:
      lC0 in riscv.o
ld: symbol(s) not found for architecture arm64
collect2: error: ld returned 1 exit status
make: *** [build/rv_histogram] Error 1

Add emulate.o, syscall.o, syscall_sdl.o, io.o, and log.o objects and
conditional build rv_async_block_clear() fixes the error and passes the
build.

ChinYikMing committed:
CI: Integrate macOS/arm64

Specify using gcc-14 instead of gcc (which typically points to the
latest version of gcc) due to a macOS limitation in the
rlalik/setup-cpp-compiler@master action. According to [1], each gcc must
have specified version.

Softfloat build warning error using gcc-14 with optimization flag O1:
In file included from src/softfloat/source/include/internals.h:42,
                 from src/softfloat/source/s_mulAddF64.c:40:
In function 'softfloat_sub128',
    inlined from 'softfloat_mulAddF64' at
src/softfloat/source/s_mulAddF64.c:185:17:
src/softfloat/source/include/primitives.h:526:17: error: 'sig128C.v64'
may be used uninitialized [-Werror=maybe-uninitialized]
  526 |     z.v64 = a64 - b64;
      |             ~~~~^~~~~
src/softfloat/source/s_mulAddF64.c: In function 'softfloat_mulAddF64':
src/softfloat/source/s_mulAddF64.c:66:20: note: 'sig128C.v64' was
declared here
   66 |     struct uint128 sig128C;
      |                    ^~~~~~~
In function 'softfloat_sub128',
    inlined from 'softfloat_mulAddF64' at
src/softfloat/source/s_mulAddF64.c:185:17:
src/softfloat/source/include/primitives.h:527:18: error: 'sig128C.v0'
may be used uninitialized [-Werror=maybe-uninitialized]
  527 |     z.v64 -= (a0 < b0);
      |              ~~~~^~~~~
src/softfloat/source/s_mulAddF64.c: In function 'softfloat_mulAddF64':
src/softfloat/source/s_mulAddF64.c:66:20: note: 'sig128C.v0' was
declared here
   66 |     struct uint128 sig128C;
Add -Wno-initialized to surpress them.

Drop the undefined behavior test on macOS/arm64 using gcc-14, as its
libsanitizer's configure.txt does not yet support it, check [2].

[1] https://github.com/rlalik/setup-cpp-compiler
[2] https://github.com/iains/gcc-darwin-arm64/blob/master-wip-apple-si/
    libsanitizer/configure.tgt

Close #519

ChinYikMing commented:
> macOS runners still suffering occasionally from fetching the latest release tag of prebuilt binaries.
>
> This is the code snippet where the macOS runners fail:
> https://github.com/sysprog21/rv32emu/blob/b8d6fa757fb65992af1e123835c0797dc9a446c8/mk/artifact.mk#L57C1-L63C10

> I planned to create three tag name which are always point to the latest prebuilt release. So that, no need to fetch the latest release tag name.
> For ELF binaries, call ELF-latest.
> For Linux images, call Linux-image-latest.
> For Sail model, call Sail-latest.
> If has any other prefer name, please let me know.
>
> Before uploading built artifact, delete the previous tag name from one of three of them by git push -d origin xxx (where xxx is one of three of the tag name). Then create the tag name and link to the release prebuilt binaries.
>
> What do you guys think?

Quote from the issue. The idea is similar to [7af92d2](https://github.com/sysprog21/rv32emu/pull/569/commits/7af92d24a09dd5368a3efaf8663aeefc60efbf80).

ChinYikMing commented:
The JIT test fails as mention in the issue. So that this changes can be postponed to be merged until the JIT warning error is handled. See [CI](https://github.com/sysprog21/rv32emu/actions/runs/13315701861/job/37189028062?pr=569).

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: ad16cd8112c43a2ba23cf32d06300636e1022826 | Previous: f6af403ed762dcf8185013ddf2148250be45b1ba | Ratio |
|-|-|-|-|
| `Dhrystone` | `1239` Average DMIPS over 10 runs | `1331` Average DMIPS over 10 runs | `1.07` |
| `Coremark` | `937.092` Average iterations/sec over 10 runs | `946.757` Average iterations/sec over 10 runs | `1.01` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

ChinYikMing commented:
The macOS host CI completes in approximately 25 minutes, exceeding the Linux host by 7 minutes but still noticeably faster than Linux/arm64. This seems acceptable for adopting the macOS host CI.

jserv commented on .github/workflows/main.yml:
```diff
@@ -286,6 +286,216 @@ jobs:
+    strategy:
+      fail-fast: false
+      matrix:
+        compiler: [gcc-14, clang]
```
I am not sure if it makes sense to validate gcc-14 builds on macOS.

ChinYikMing replied:
> I am not sure if it makes sense to validate gcc-14 builds on macOS.

gcc is still available via `brew install gcc` (specifically as gcc-14, the latest version). Embracing gcc builds in CI on macOS could benefit gcc enthusiasts? Additionally, building with gcc on macOS uncovered some potential build errors or warnings, which have been addressed in these changes.

jserv replied:
I agree with the use of gcc. Let's skip the tests on different optimization flags since we already check the general optimizations on Ubuntu Linux prior to macOS builds.

ChinYikMing replied:
> I agree with the use of gcc. Let's skip the tests on different optimization flags since we already check the general optimizations on Ubuntu Linux prior to macOS builds.

But, the softfloat warning is revealed with `gcc-14` + O1 flag on macOS, see the commit message of [1d497b1](https://github.com/sysprog21/rv32emu/pull/569/commits/1d497b102a6fb5b5dd0bd8a33af6b3527466e475). Shall we keep them?

jserv replied:
> But, the softfloat warning is revealed with `gcc-14` + O1 flag on macOS, see the commit message of [1d497b1](https://github.com/sysprog21/rv32emu/pull/569/commits/1d497b102a6fb5b5dd0bd8a33af6b3527466e475). Shall we keep them?

The default `CFLAGS` includes `-O2`, and it should be sufficient to make the compilation optimization issues verbose.

ChinYikMing replied:
> > But, the softfloat warning is revealed with `gcc-14` + O1 flag on macOS, see the commit message of [1d497b1](https://github.com/sysprog21/rv32emu/pull/569/commits/1d497b102a6fb5b5dd0bd8a33af6b3527466e475). Shall we keep them?
> 
> The default `CFLAGS` includes `-O2`, and it should be sufficient to make the compilation optimization issues verbose.

Got it, but I will remain the '-Wno-uninitialized' CFLAGS for softfloat.

jserv commented on .github/workflows/main.yml:
```diff
@@ -286,6 +286,118 @@ jobs:
+          .ci/riscv-toolchain-install.sh && export PATH=$PWD/toolchain/bin:$PATH
+          python3 -m venv venv
+          . venv/bin/activate
+          .ci/riscv-tests.sh
```
commit 9e775bacdd83012137a9aa30eca08216642d33d0 drops unstable architecture test on Arm64/Linux. Remove the RISC-V Architecture Tests here.

ChinYikMing commented:
TODO: add emcc build in CI. #576

Edit: done.

ChinYikMing committed:
CI: Add emcc build on macOS/arm64

ChinYikMing committed:
CI: Shorten the statement with PARALLEL variable

ChinYikMing committed:
CI: Fix failure when fetching latest release tag on macOS/arm64

Refer to [1] and [2], leveraging ${{ secrets.GITHUB_TOKEN }} to be
authenticated which increases the Github API request rate limit.

[1] https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#primary-rate-limit-for-github_token-in-github-actions
[2] https://docs.github.com/en/rest/using-the-rest-api/rate-limits-for-the-rest-api?apiVersion=2022-11-28#getting-a-higher-rate-limit

vacantron commented on .github/workflows/main.yml:
```diff
@@ -287,6 +290,9 @@ jobs:
       uses: actions/checkout@v4
+    - name: Set parallel jobs variable
+      run: |
+            echo "PARALLEL=-j$(nproc)" >> $GITHUB_ENV
```
Because `run-on-arch-action` runs in Docker and QEMU, the enviornment variable `$GITHUB_*` could not be able to access in container and it will complain about unbound variable.

ChinYikMing replied:
> Because `run-on-arch-action` runs in Docker and QEMU, the enviornment variable `$GITHUB_*` could not be able to access in container and it will complain about unbound variable.

Oh yes, it could be solved by actually passing the PARALLEL to the environment (trying this).
https://github.com/sysprog21/rv32emu/blob/6928f8ccd4257c2020f1a261dea2d7354d9cafee/.github/workflows/main.yml#L308

```diff
...
        env: |
          CC: clang-18
+.        PARALLEL: ${{ env.PARALLEL }}
...
```

vacantron committed:
jit: Suppress compiler error on macOS/arm64

unused function 'emit_dataproc_3source' [-Werror,-Wunused-function]

ChinYikMing commented:
> The JIT test fails as mention in the issue. So that this changes can be postponed to be merged until the JIT warning error is handled. See [CI](https://github.com/sysprog21/rv32emu/actions/runs/13315701861/job/37189028062?pr=569).

Fixed in commit 297db9a, check [CI](https://github.com/sysprog21/rv32emu/actions/runs/13574305343).

ChinYikMing commented:
One remaining issue to be fixed is that fetching failure of the latest release tag of rv32emu-prebuilt. on macOS/arm64, the failure can be happened very often. The commit da51ade eliminates calling the Github API then avoid the issue. Trying to investigate the root cause.

jserv commented:
Thank @ChinYikMing for contributing!
